akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -757,11 +786,38 @@ abstract class ShardCoordinator(
               self.tell(GetShardHome(shard), ignoreRef)
             }
           } else {
-            // rebalance not completed, graceful shutdown will be retried
-            gracefulShutdownInProgress -= state.shards(shard)
             clearRebalanceInProgress(shard)
           }
         } else {
+          log.warning("Rebalance shard [{}] didn't complete within [{}].", shard, handOffTimeout.pretty)
+          clearRebalanceInProgress(shard)
+        }
+
+      case ShardShutdownDone(shard, ok) =>
+        rebalanceWorkers -= sender()
+
+        if (ok) {
+          log.debug("Shard [{}] shutdown completed successfully.", shard)
+          if (state.shards.contains(shard)) {
+            update(ShardHomeDeallocated(shard)) { evt =>
+              log.debug("Shard [{}] deallocated after after region shutdown", shard)
+              state = state.updated(evt)
+              clearRebalanceInProgress(shard)
+              allocateShardHomesForRememberEntities()
+              self.tell(GetShardHome(shard), ignoreRef)
+            }
+          } else {
+            clearRebalanceInProgress(shard)
+          }
+        } else {
+          log.warning("Shard [{}] shutdown didn't complete within [{}].", shard, handOffTimeout.pretty)
+          // TODO: we should take that into consideration for future rebalances
+          // the region is trying to shutdown but hangs because of a shard,
+          // we may want to not allocate new shards on this region
```

Comment:
I think that is handled (at least after a while) because we avoid allocating to leaving members. It's possible that we haven't seen the leaving status here yet, but I'm not sure it's worth keeping track of another "in progress" thing.